wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -2515,4 +2449,430 @@ private class WasmExpressionBuilder private (
+
+          val nextTryFinallyEntry = innermostTryFinally // note that we're out of ourselves already
+            .filter(nextTry => possibleTargetEntries.exists(nextTry.isInside(_)))
+
+          /* Build the destination table for `br_table`. Target Labeled's that
+           * are outside of the next try..finally in line go to the latter;
+           * for other `Labeled`'s, we go to their cross label.
+           */
+          val brTableDests: List[(Int, WasmLabelName)] = possibleTargetEntries.map { targetEntry =>
+            val (destinationTag, _, crossLabel) = targetEntry.requireCrossInfo()
+            val label = nextTryFinallyEntry.filter(_.isInside(targetEntry)) match {
+              case None          => crossLabel
+              case Some(nextTry) => nextTry.requireCrossInfo()._2
+            }
+            destinationTag -> label
+          }
+
+          instrs += LOCAL_GET(entry.requireCrossInfo()._1)
+          for (nextTry <- nextTryFinallyEntry) {
+            // Transfer the destinationTag to the next try..finally in line
+            instrs += LOCAL_TEE(nextTry.requireCrossInfo()._1)
+          }
+          emitBRTable(brTableDests, doneLabel)
+        }
+      } // end block $done
+
+      // reload the result onto the stack
+      for (resultLocal <- resultLocals)
+        instrs += LOCAL_GET(resultLocal)
+
+      if (t.tpe == IRTypes.NothingType)
+        instrs += UNREACHABLE
+
+      t.tpe
+    }
+
+    private def emitBRTable(
+        dests: List[(Int, WasmLabelName)],
+        defaultLabel: WasmLabelName
+    ): Unit = {
+      dests match {
+        case Nil =>
+          instrs += DROP
+          instrs += BR(defaultLabel)
+
+        case (singleDestValue, singleDestLabel) :: Nil =>
+          /* Common case (as far as getting here in the first place is concerned):
+           * All the `Return`s that cross the current `TryFinally` have the same
+           * target destination (namely the enclosing `def` in the original program).
+           */
+          instrs += I32_CONST(singleDestValue)
+          instrs += I32_EQ
+          instrs += BR_IF(singleDestLabel)
+          instrs += BR(defaultLabel)
+
+        case _ :: _ =>
+          // `max` is safe here because the list is non-empty
+          val table = Array.fill(dests.map(_._1).max + 1)(defaultLabel)
+          for (dest <- dests)
+            table(dest._1) = dest._2
+          instrs += BR_TABLE(table.toList, defaultLabel)
+      }
+    }
+
+    def genReturn(t: IRTrees.Return): IRTypes.Type = {
+      val targetEntry = enclosingLabeledBlocks(t.label.name)
+
+      genTree(t.expr, targetEntry.expectedType)
+
+      if (targetEntry.expectedType != IRTypes.NothingType) {
+        innermostTryFinally.filter(_.isInside(targetEntry)) match {
+          case None =>
+            // Easy case: directly branch out of the block
+            instrs += BR(targetEntry.regularWasmLabel)
```

Comment:
Scala 3's boundary/break is indeed compiled to `Labeled/Return`. However the optimization is explicitly not allowed if there is a `try..finally` in the middle of the `boundary/break`. In that case it falls back to actually throwing and catching an exception. That's because the JVM backend cannot deal with it. The JVM backend only handles `return` in that situation.

We have to deal with it because a `return` keyword in Scala source is compiled as a `Labeled/Return` as well. After inlining (once we enable the optimizer), we can truly get arbitrary combinations.